tests/test_batch_run.py

```diff
@@ -85,6 +85,7 @@ def get_local_model_param(self):
     def step(self):
         self.datacollector.collect(self)
         self.schedule.step()
+        self.advance_time()
```

Comment:
If we can remove this line from being needed, we are fully compliant with current user methods. I would see `deltat` as an input argument of step itself (defaulting to one).

Reply:
My take on the alternative was:

> I had thought about super().step(). The problem is that it is very easy to forget doing so. It doesn't have a direct meaning of incrementing the time.

I suppose it is the choice being remembering to always:
- call super on an inherited method
- advance the clock and steps

I pick the latter because it is more explicit on what it does. But I can see the former being lower on the LOC needed, because I can also sweep `self.datacolllector.collect(self) if hasattr(self, "datacollector") else None` into `super().step()`.

Reply:
After writing it more explicitly,
```python
def step(self, deltat=1):
    self.agents.shuffle().do("step")
    super().step(deltat)
```
One problem is that the time update needs to happen last. And the super call being called last doesn't seem conventional.

Reply:
Could we use a decorator for time advancement?

```Python
def time_advancement_decorator(step_func):
    def wrapper(self, *args, **kwargs):
        # Extract deltat from kwargs or default to 1 if not provided
        deltat = kwargs.get('deltat', 1)
        step_func(self, *args, **kwargs)
        self.advance_time(deltat)
    return wrapper

class Model:
    @time_advancement_decorator
    def step(self, deltat=1):
        # Base step implementation
        pass

class MyModel(Model):
    def step(self):
        # Subclass-specific logic
        pass
```

Reply:
That should do. It frees the user from having to deal with the location of the `advance_time` within the `step()` specification (that it should be called last). I'm in favor of using a decorator.

Reply:
I didn't see the 2 models in the code. What I had in mind was to specify the decorator in `MyModel`.
```python
class MyModel(Model):
    @time_advancement_decorator
    def step(self):
        # Subclass-specific logic
        pass
```

Reply:
> That should do. It frees the user from having to deal with the location of the `advance_time` within the `step()` specification (that it should be called last). I'm in favor of using a decorator.

Just for my understanding, why should time be advanced last?

Reply:
I suppose it would be fine to advance time first before everything else, at the expense of the steps starting from 1, and the time starting from `deltat`, instead of 0.

Reply:
The challenge here is to find a solution which requires no changes to existing user models. Otherwise we have to bump to 3.0.

Reply:
> The challenge here is to find a solution which requires no changes to existing user models

That is easy: leave the schedulers alone for now for backward compatibility while slowly putting in place more sophisticated machinery (as we did with the implementation of the scheduler on top of AgentSet). 

> I think it's just convention to maintain roughly this order:

> Do things
Collect data
Increase time

Interesting. For me, the obvious order is increment time, do stuff, and gather data. But indeed, It does not really matter.

Reply:
Yeah that's also an obvious one. If you then collect data once after initialisation you also capture the initial model state at timestep 0, and the final model state at timestep n, giving a database of n+1 length.

Maybe we should formalize a good default for this somewhere.

(note that the order of do stuff and increase time not matter, only collect data does)

Reply:
I think we have this resolved with the wrapper step right?

Reply:
Marking as resolved, we solved this with the wrapper idea in https://github.com/projectmesa/mesa/pull/1942#issuecomment-1903419315.